cmd/zetaclientd-supervisor/lib.go

```diff
@@ -0,0 +1,258 @@
+	default:
+		logger = zerolog.New(zerolog.ConsoleWriter{Out: out, TimeFormat: time.RFC3339})
+	}
+
+	return logger
+}
+
+type zetaclientdSupervisor struct {
+	zetacoredConn   *grpc.ClientConn
+	reloadSignals   chan bool
+	logger          zerolog.Logger
+	upgradesDir     string
+	upgradePlanName string
+}
+
+func newZetaclientdSupervisor(zetaCoreURL string, logger zerolog.Logger) (*zetaclientdSupervisor, error) {
+	logger = logger.With().Str("module", "zetaclientdSupervisor").Logger()
+	conn, err := grpc.Dial(
+		fmt.Sprintf("%s:9090", zetaCoreURL),
+		grpc.WithInsecure(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("grpc dial: %w", err)
+	}
+
+	return &zetaclientdSupervisor{
+		zetacoredConn: conn,
+		logger:        logger,
+		reloadSignals: make(chan bool, 1),
+		upgradesDir:   os.ExpandEnv("$HOME/.zetaclientd/upgrades"),
+	}, nil
+}
+
+func (s *zetaclientdSupervisor) Start(ctx context.Context) {
+	go s.watchForVersionChanges(ctx)
+	go s.handleCoreUpgradePlan(ctx)
+}
+
+func (s *zetaclientdSupervisor) WaitForReloadSignal(ctx context.Context) {
+	select {
+	case <-s.reloadSignals:
+	case <-ctx.Done():
+	}
+}
+
+func (s *zetaclientdSupervisor) dirForVersion(version string) string {
+	return path.Join(s.upgradesDir, version)
+}
+
+func atomicSymlink(target, linkName string) error {
+	linkNameTmp := linkName + ".tmp"
+	_, err := os.Stat(target)
+	if err != nil {
+		return fmt.Errorf("stat target: %w", err)
+	}
+	err = os.Remove(linkNameTmp)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("remove old current tmp: %w", err)
+	}
+	err = os.Symlink(target, linkNameTmp)
+	if err != nil {
+		return fmt.Errorf("new symlink: %w", err)
+	}
+	err = os.Rename(linkNameTmp, linkName)
+	if err != nil {
+		return fmt.Errorf("rename symlink: %w", err)
+	}
+	return nil
+}
+
+func (s *zetaclientdSupervisor) watchForVersionChanges(ctx context.Context) {
+	client := tmservice.NewServiceClient(s.zetacoredConn)
+	prevVersion := ""
+	for {
+		select {
+		case <-time.After(time.Second):
+		case <-ctx.Done():
+			return
+		}
+		res, err := client.GetNodeInfo(ctx, &tmservice.GetNodeInfoRequest{})
+		if err != nil {
+			s.logger.Warn().Err(err).Msg("get node info")
+			continue
+		}
+		newVersion := res.ApplicationVersion.Version
+		if prevVersion == "" {
+			prevVersion = newVersion
+		}
+		if prevVersion == newVersion {
+			continue
+		}
+		s.logger.Warn().Msgf("core version change (%s -> %s)", prevVersion, newVersion)
+
+		prevVersion = newVersion
+
+		// TODO: just use newVersion when #2135 is merged
+		// even without #2135, the version will still change and trigger the update
+		newVersionDir := s.dirForVersion(s.upgradePlanName)
+		currentLinkPath := s.dirForVersion("current")
+
+		err = atomicSymlink(newVersionDir, currentLinkPath)
+		if err != nil {
+			s.logger.Error().Err(err).Msgf("unable to update current symlink (%s -> %s)", newVersionDir, currentLinkPath)
+			return
+		}
+		s.reloadSignals <- true
```

Comment:
Would this trigger in the block after the upgrade happens on zetacore ? 

We can maybe consider saving the upgrade height as well during handleCoreUpgrade , so that the clients can be restart at the height directly , wdyt

Reply:
This will not trigger until the zetacored upgrade has completed.

zetaclient itself should detect once upgrade height has been reached and pause itself:

https://github.com/zeta-chain/node/blob/e0287d7d784c2f20c637e8d0409704a3a768cdf3/zetaclient/zetabridge/zetacore_bridge.go#L196-L200

although if blocks are created faster than the the core context updater ticker then zetaclient wouldn't actually pause itself (I'm thinking about mainnet)